apps/www/_blog/2023-07-13-pgvector-performance.mdx

```diff
@@ -0,0 +1,297 @@
+
+```sql
+SET maintenance_work_mem TO '7168 MB';
+```
+
+</div>
+
+Keeping in mind that the overall index size is almost the same, and only index build time increases, we can say that it's better to use more lists for better select queries speed.
+
+### Real data has higher precision than random data
+
+Embeddings created from “real” data are more likely to be clustered together, whereas random embeddings are more likely to be scattered. In other words, real embeddings are very far from being randomly distributed. This might seem obvious, but it's an important call-out for benchmarks.
+
+Embeddings generated for similarity search using “real world data” will be more correlated, so the precision will be higher as well. You can see the difference in this chart using 10,000 Wikipedia embeddings, vs 10,000 randomly-generated embeddings:
+
+<div>
+  <img
+    alt="multi database"
+    className="dark:hidden"
+    src="/images/blog/2023-07-13-pgvector-performance/random-data--light.png"
+  />
+  <img
+    alt="multi database"
+    className="hidden dark:block"
+    src="/images/blog/2023-07-13-pgvector-performance/random-data--dark.png"
+  />
+</div>
+
+## Optimizing pgvector performance
+
+Armed with all this information, we can safely give you a few tips and strategies for optimizing your pgvector workloads.
+
+### Tips
+
+First, a few generic tips which you can pick and choose from:
+
+1. **Adjust your Postgres config.** It should be aligned with RAM & CPU cores. [Find more details here](https://gist.github.com/egor-romanov/323e2847851bbd758081511785573c08).
+2. Prefer `inner-product` to `L2` or `Cosine` distances if your vectors are normalized (like `text-embedding-ada-002`). If embeddings are not normalized, `Cosine` distance should give the best results with an index.
+3. **Pre-warm your database.** Implement the warm-up technique we described earlier before transitioning to production.
+4. **Establish your workload.** Increasing the lists constant for the pgvector index can accelerate your queries (at the expense of a slower build). For instance, for benchmarks with OpenAI embeddings, we employed a `lists` constant of 2000 (`number of vectors / 500`) as opposed to the suggested 1000 (`number of vectors / 1000`).
+5. **Benchmark your own specific workloads.** Doing this during cache warm-up helps gauge the best value for the `probes` constant, balancing precision with RPS.
+
+### Going into production
+
+Before running your pgvector workload in production, here are a few steps you can take to maximize performance.
+
+1. Over-provision RAM during preparation. You can scale down in step `5`, but it's better to start with a larger size to get the best results for RAM requirements. (We'd recommend at least 8XL if you're using Supabase.)
+2. Upload your data to the database. If you use `[vecs](https://supabase.com/docs/guides/ai/python/api)` library, it will automatically generate an index with default parameters.
```

Comment:
```suggestion
2. Upload your data to the database. If you use [`vecs`](https://supabase.com/docs/guides/ai/python/api) library, it will automatically generate an index with default parameters.
```